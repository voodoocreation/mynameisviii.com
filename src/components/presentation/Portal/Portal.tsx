import * as React from "react";
import * as ReactDOM from "react-dom";

import { isServer } from "../../../helpers/dom";

interface IProps {
  className?: string;
  isRenderingInPlace?: boolean;
}

export default class Portal extends React.Component<IProps> {
  public static defaultProps = {
    isRenderingInPlace: false
  };

  private container: HTMLDivElement | undefined = !isServer()
    ? document.createElement("div")
    : undefined;

  public componentDidMount() {
    if (!isServer() && this.container) {
      this.container.className = `${this.props.className}`;

      const portal = this.getPortalNode();
      portal.appendChild(this.container);
    }
  }

  public componentWillUnmount() {
    if (!isServer() && this.container) {
      const portal = this.getPortalNode();
      portal.removeChild(this.container);
    }
  }

  public render() {
    if (!this.container || this.props.isRenderingInPlace) {
      return <div className="Portal--inPlace">{this.props.children}</div>;
    }

    return ReactDOM.createPortal(this.props.children, this.container);
  }

  private getPortalNode = () => {
    let portal = document.querySelector(".Portal");

    if (!portal) {
      portal = document.createElement("div");
      portal.className = "Portal";
      document.body.appendChild(portal);
    }

    return portal;
  };
}
