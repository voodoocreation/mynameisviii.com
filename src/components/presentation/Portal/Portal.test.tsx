import * as React from "react";
import { Wrapper } from "react-test-wrapper";

import Portal from "./Portal";

const component = new Wrapper(Portal)
  .withDefaultProps({ className: "TestPortal" })
  .withDefaultChildren(<div className="TestPortalComponent" />);

describe("[presentation] <Portal />", () => {
  describe("when mounting on the server", () => {
    Object.defineProperty(window, "isServer", {
      value: true,
      writable: true
    });

    const wrapper = component.mount();

    it("renders in-place", () => {
      expect(
        wrapper.find(".Portal--inPlace .TestPortalComponent")
      ).toHaveLength(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });

    it("unmounts component", () => {
      wrapper.unmount();
    });
  });

  describe("when mounting on the client", () => {
    Object.defineProperty(window, "isServer", {
      value: false,
      writable: true
    });

    const wrapper = component.mount();

    it("creates the portal node", () => {
      expect(document.querySelectorAll(".Portal")).toHaveLength(1);
    });

    it("creates the instance container with the defined class", () => {
      expect(document.querySelectorAll(".Portal .TestPortal")).toHaveLength(1);
    });

    it("doesn't render in-place", () => {
      expect(wrapper.find(".Portal--inPlace")).toHaveLength(0);
    });

    it("renders the children", () => {
      expect(wrapper.find(".TestPortalComponent")).toHaveLength(1);
    });

    it("matches snapshot", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });

    it("unmounts component", () => {
      wrapper.unmount();
    });

    it("removes the instance container from the DOM", () => {
      expect(document.querySelectorAll(".Portal *")).toHaveLength(0);
    });
  });
});
