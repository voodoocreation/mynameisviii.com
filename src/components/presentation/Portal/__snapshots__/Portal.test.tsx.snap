// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Portal /> when mounting on the client matches snapshot 1`] = `
<div
  class="TestPortalComponent"
/>
`;

exports[`[presentation] <Portal /> when mounting on the server matches snapshot 1`] = `
<div
  class="Portal--inPlace"
>
  <div
    class="TestPortalComponent"
  />
</div>
`;
